From 1c5644ef585ed8ac124eae00a358912dca703421 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Mon, 16 Jul 2018 10:39:31 +0200
Subject: [PATCH] Recorder: Replace hardcoded media_profiles path with a list
 of possible paths

* This fixes invalid resolution of screen recording for devices
  where media_profiles.xml isn't in /system/etc.

Change-Id: Icf02b8f915db8ab0edd58678b40a74c42cf1fb1a
---
 .../recorder/screen/EncoderDevice.java        | 25 ++++++++++++++++++-
 1 file changed, 24 insertions(+), 1 deletion(-)

diff --git a/app/src/main/java/org/lineageos/recorder/screen/EncoderDevice.java b/app/src/main/java/org/lineageos/recorder/screen/EncoderDevice.java
index e4c5ff4..a9639f9 100644
--- a/app/src/main/java/org/lineageos/recorder/screen/EncoderDevice.java
+++ b/app/src/main/java/org/lineageos/recorder/screen/EncoderDevice.java
@@ -23,6 +23,7 @@ import android.media.CamcorderProfile;
 import android.media.MediaCodec;
 import android.media.MediaCodecInfo;
 import android.media.MediaFormat;
+import android.os.SystemProperties;
 import android.text.TextUtils;
 import android.util.Log;
 import android.view.Surface;
@@ -36,6 +37,7 @@ import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.StringReader;
 import java.util.ArrayList;
+import java.util.List;
 
 import safesax.Element;
 import safesax.ElementListener;
@@ -77,6 +79,13 @@ abstract class EncoderDevice {
             //{ 960, 540 },
             {848, 480}
     };
+    private List<String> validMediaProfilesPaths = new ArrayList<String>() {{
+            add("/system/etc/media_profiles.xml");
+            add("/system/etc/media_profiles_vendor.xml");
+            add("/vendor/etc/media_profiles.xml");
+            add("/vendor/etc/media_profiles_V1_0.xml");
+            add("/vendor/etc/media_profiles_vendor.xml");
+    }};
     private MediaCodec venc;
     private int width;
     private int height;
@@ -149,12 +158,26 @@ abstract class EncoderDevice {
             venc = null;
         }
 
+        String mediaProfilesPath = "";
+        String mediaProfilesProp = SystemProperties.get("media.settings.xml", "");
         int maxWidth;
         int maxHeight;
         int bitrate;
 
+        if (!mediaProfilesProp.isEmpty()) {
+            validMediaProfilesPaths.add(0, mediaProfilesProp);
+        }
+
+        for (String path : validMediaProfilesPaths) {
+            File mediaProfiles = new File(path);
+            if (mediaProfiles.canRead()) {
+                mediaProfilesPath = path;
+                break;
+            }
+        }
+
         try {
-            File mediaProfiles = new File("/system/etc/media_profiles.xml");
+            File mediaProfiles = new File(mediaProfilesPath);
             FileInputStream fin = new FileInputStream(mediaProfiles);
             byte[] bytes = new byte[(int) mediaProfiles.length()];
             //noinspection ResultOfMethodCallIgnored
-- 
2.17.1

